### PR TITLE
Adds LogCumSumExp

### LINKED ISSUE #26411: LogCumsumExp
## 🚀 Feature
Add numerically stable cumulative logsumexp function. Also we have associated PR on `cummax` that is needed for numerically stable implementation (https://github.com/pytorch/pytorch/issues/20240).
## Motivation
This is useful when computing sum of probabilities and have different applications.

## Pitch
Torch has `cumsum` and `cumprod` so I suggest `logcumsumexp` to be added.

## Alternatives
Current workaround for me and for people that need it can be written as follows (it is quite fast, only overhead for converting between numpy/torch):
```
import numpy as np
import torch

def cummax(x, dim):
    x_np = x.detach().cpu().numpy()
    ret = np.maximum.accumulate(x_np, axis=dim)
    return torch.from_numpy(ret).to(x)


def logcumsumexp(x, dim=-1):
    if (dim != -1) or (dim != x.ndimension() - 1):
        x = x.transpose(dim, -1)

    init_size = x.size()
    last_dim_size = init_size[-1]
    x_resized = x.contiguous().view(-1, last_dim_size)
    d1, d2 = x_resized.size()
    x_cummax = cummax(x_resized, -1).view(d1, d2, 1)
    x_expand = x_resized.unsqueeze(1).expand(d1, d2, last_dim_size)
    mask = torch.tril(torch.ones(last_dim_size, last_dim_size)).unsqueeze(0)
    ret = torch.log(torch.sum(torch.exp(x_expand - x_cummax) * mask, dim=-1)) + x_cummax.view(d1, d2)
    ret = ret.view(*init_size)

    if (dim != -1) or (dim != x.ndimension() - 1):
        ret = ret.transpose(-1, dim)

    return ret
```

UPDATE: code above is not always numerically stable. Still most numerically stable, but slow way to compute `logcumsumexp` is use for-loop:

```
def logcumsumexp(x, dim):
    # slow implementation, but ok for now
    if (dim != -1) or (dim != x.ndimension() - 1):
        x = x.transpose(dim, -1)

    out = []
    for i in range(1, x.size(-1) + 1):
        out.append(torch.logsumexp(x[..., :i], dim=-1, keepdim=True))
    out = torch.cat(out, dim=-1)

    if (dim != -1) or (dim != x.ndimension() - 1):
        out = out.transpose(-1, dim)
    return out
```

### PR DESCRIPTION
**Pull request by pandeykartikey**
Resolves #26411

### CODE REVIEW

**Comment by pandeykartikey** [comment]
@agadetsky Have a look.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit cb6368b8:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=cb6368b83ced5d56df75faf410e33085f248ce8b).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4391506) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93858892) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93858892))
<details>
<summary>
<code>Jan 31 20:50:48 RuntimeError: could not find ATen declaration for schema: cumlogsumexp(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor .  Available signatures:</code>
</summary>

```
Jan 31 20:50:48   File "tools/setup_helpers/generate_code.py", line 43, in generate_code 
Jan 31 20:50:48     gen_autograd_python(declarations_path or DECLARATIONS_PATH, autograd_gen_dir, 'tools/autograd') 
Jan 31 20:50:48   File "/var/lib/jenkins/workspace/tools/autograd/gen_autograd.py", line 219, in gen_autograd_python 
Jan 31 20:50:48     os.path.join(autograd_dir, 'derivatives.yaml'), aten_decls) 
Jan 31 20:50:48   File "/var/lib/jenkins/workspace/tools/autograd/load_derivatives.py", line 28, in load_derivatives 
Jan 31 20:50:48     for defn in definitions] 
Jan 31 20:50:48   File "/var/lib/jenkins/workspace/tools/autograd/load_derivatives.py", line 28, in <listcomp> 
Jan 31 20:50:48     for defn in definitions] 
Jan 31 20:50:48   File "/var/lib/jenkins/workspace/tools/autograd/load_derivatives.py", line 197, in process_definition 
Jan 31 20:50:48     '.  Available signatures:\n{}'.format(specification, '\n'.join(avail))) 
Jan 31 20:50:48 RuntimeError: could not find ATen declaration for schema: cumlogsumexp(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor .  Available signatures: 
Jan 31 20:50:48 cumlogsumexp(Tensor self, int dim) -> Tensor 
Jan 31 20:50:48 [ 34%] Building C object confu-deps/cpuinfo/CMakeFiles/cpuinfo.dir/src/x86/uarch.c.o 
Jan 31 20:50:48 caffe2/CMakeFiles/generate-torch-sources.dir/build.make:107: recipe for target '../torch/csrc/autograd/generated/Functions.cpp' failed 
Jan 31 20:50:48 make[2]: *** [../torch/csrc/autograd/generated/Functions.cpp] Error 1 
Jan 31 20:50:48 CMakeFiles/Makefile2:5482: recipe for target 'caffe2/CMakeFiles/generate-torch-sources.dir/all' failed 
Jan 31 20:50:48 make[1]: *** [caffe2/CMakeFiles/generate-torch-sources.dir/all] Error 2 
Jan 31 20:50:48 make[1]: *** Waiting for unfinished jobs.... 
Jan 31 20:50:48 [ 34%] Building C object confu-deps/cpuinfo/CMakeFiles/cpuinfo.dir/src/x86/name.c.o 
Jan 31 20:50:49 [ 34%] Building C object confu-deps/cpuinfo/CMakeFiles/cpuinfo.dir/src/x86/topology.c.o 
Jan 31 20:50:49 [ 34%] Building C object confu-deps/cpuinfo/CMakeFiles/cpuinfo.dir/src/x86/isa.c.o 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4391509) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93858921) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93858921))
<details>
<summary>
<code>Jan 31 20:50:44 RuntimeError: could not find ATen declaration for schema: cumlogsumexp(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor .  Available signatures:</code>
</summary>

```
Jan 31 20:50:44   File "tools/setup_helpers/generate_code.py", line 88, in main 
Jan 31 20:50:44     options.selected_op_list_path, 
Jan 31 20:50:44   File "tools/setup_helpers/generate_code.py", line 43, in generate_code 
Jan 31 20:50:44     gen_autograd_python(declarations_path or DECLARATIONS_PATH, autograd_gen_dir, 'tools/autograd') 
Jan 31 20:50:44   File "/var/lib/jenkins/workspace/tools/autograd/gen_autograd.py", line 219, in gen_autograd_python 
Jan 31 20:50:44     os.path.join(autograd_dir, 'derivatives.yaml'), aten_decls) 
Jan 31 20:50:44   File "/var/lib/jenkins/workspace/tools/autograd/load_derivatives.py", line 28, in load_derivatives 
Jan 31 20:50:44     for defn in definitions] 
Jan 31 20:50:44   File "/var/lib/jenkins/workspace/tools/autograd/load_derivatives.py", line 197, in process_definition 
Jan 31 20:50:44     '.  Available signatures:\n{}'.format(specification, '\n'.join(avail))) 
Jan 31 20:50:44 RuntimeError: could not find ATen declaration for schema: cumlogsumexp(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor .  Available signatures: 
Jan 31 20:50:44 cumlogsumexp(Tensor self, int dim) -> Tensor 
Jan 31 20:50:44 make[2]: *** [../torch/csrc/autograd/generated/Functions.cpp] Error 1 
Jan 31 20:50:44 caffe2/CMakeFiles/generate-torch-sources.dir/build.make:107: recipe for target '../torch/csrc/autograd/generated/Functions.cpp' failed 
Jan 31 20:50:44 make[1]: *** [caffe2/CMakeFiles/generate-torch-sources.dir/all] Error 2 
Jan 31 20:50:44 make[1]: *** Waiting for unfinished jobs.... 
Jan 31 20:50:44 CMakeFiles/Makefile2:5524: recipe for target 'caffe2/CMakeFiles/generate-torch-sources.dir/all' failed 
Jan 31 20:50:44 [ 35%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/java/java_string_field_lite.cc.o 
Jan 31 20:50:45 [ 35%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/js/js_generator.cc.o 
Jan 31 20:50:45 [ 35%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/js/well_known_types_embed.cc.o 
Jan 31 20:50:45 [ 35%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_enum.cc.o 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4391523) pytorch_linux_xenial_py3_clang5_mobile_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93859569) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93859569))
<details>
<summary>
<code>Jan 31 20:50:57 caused by: Connection refused (os error 111)</code>
</summary>

```
Jan 31 20:50:57 +++ eval 'extract_trap_cmd ' 
Jan 31 20:50:57 ++++ extract_trap_cmd 
Jan 31 20:50:57 ++++ printf '%s\n' '' 
Jan 31 20:50:57 +++ printf '%s\n' cleanup 
Jan 31 20:50:57 ++ trap -- ' 
Jan 31 20:50:57 cleanup' EXIT 
Jan 31 20:50:57 ++ which sccache 
Jan 31 20:50:57 ++ sccache --stop-server 
Jan 31 20:50:57 Stopping sccache server... 
Jan 31 20:50:57 error: couldn't connect to server 
Jan 31 20:50:57 caused by: Connection refused (os error 111) 
Jan 31 20:50:57 ++ true 
Jan 31 20:50:57 ++ rm /var/lib/jenkins/sccache_error.log 
Jan 31 20:50:57 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Jan 31 20:50:57 ++ true 
Jan 31 20:50:57 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Jan 31 20:50:57 ++ SCCACHE_IDLE_TIMEOUT=1200 
Jan 31 20:50:57 ++ RUST_LOG=sccache::server=error 
Jan 31 20:50:57 ++ sccache --start-server 
Jan 31 20:50:57 Starting sccache server... 
Jan 31 20:50:57 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/test_torch.py [hunk 1]
```diff
@@ -10013,6 +10013,22 @@ def test_ops(op, string_of_function_name, expected_output):
...
                                                        [0, 0, 0]]))
 
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_cumlogsumexp(self):
+        a = torch.randn(5, 4)
+        a[0, 0] = inf
+        a[1, :] = -inf
+        actual = a.cumlogsumexp(1)
+        max = np.maximum.accumulate(a.numpy(), axis=1)
+        expected = np.log(max + np.cumsum(np.exp(a - max), axis=1))
```
Comment (agadetsky) [inline test/test_torch.py]: addition of `max` must be after `log`.

Comment (pandeykartikey) [reply test/test_torch.py]: Fixed. Have a look.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -180,6 +180,25 @@ static TensorIterator make_reduction(
...
+  check_scalar_type_device_layout_equal(result, self);
+  {
+    NoNamesGuard guard;
+    result.resize_(self.sizes());
+    auto cummax_values = std::get<0>(at::cummax(self, dim));
+    result = at::exp(self - cummax_values);
+    at::_cumsum(result, dim);
+    result.log_().add_(cummax_values);
+  }
+  return result;
```
Comment (anjali411) [inline aten/src/ATen/native/ReduceOps.cpp]: I think you should propagate names before returning result
```namedinference::propagate_names(result, self);```

**Comment by anjali411** [comment]
1. You should add the following tests: 1. test in test/test_named_tensor.py to ensure the dimension names are correctly propagated. eg. https://github.com/pytorch/pytorch/blob/10183061eb262fba98dcc2d76f6d31765609bcaf/test/test_namedtensor.py#L980
2. neg_dim_test, zero dim check(test_scalar_check),  test_dim_function_empty (meant for functions that operate over a dimension but don't reduce), tensor_op_test to compare the tensor op results on CPU and device in test_torch.py

File: test/test_torch.py [hunk 3]
```diff
@@ -10013,6 +10013,22 @@ def test_ops(op, string_of_function_name, expected_output):
...
+        max = np.maximum.accumulate(a.numpy(), axis=1)
+        expected = max + np.log(np.cumsum(np.exp(a - max), axis=1))
+        self.assertEqual(expected.shape, actual.shape)
+        self.assertTrue(np.allclose(expected, actual.numpy()))
+        # check that out is actually inplace
+        b = torch.zeros(5, 2)
+        c = b[:, 0]
+        torch.cumlogsumexp(a, 1, out=c)
+        self.assertTrue(np.allclose(expected, b[:, 0].numpy()))
+
```
Comment (anjali411) [inline test/test_torch.py]: add a test to ensure that there's a runtime error when the input and output doesn't have the same dtype

**Review by anjali411 — CHANGES_REQUESTED** [review]
1. You should add the following tests: 1. test in test/test_named_tensor.py to ensure the dimension names are correctly propagated. eg. https://github.com/pytorch/pytorch/blob/10183061eb262fba98dcc2d76f6d31765609bcaf/test/test_namedtensor.py#L980
2. neg_dim_test, zero dim check(test_scalar_check),  test_dim_function_empty (meant for functions that operate over a dimension but don't reduce), tensor_op_test to compare the tensor op results on CPU and device.
in test_torch.py

**Comment by pandeykartikey** [comment]
@anjali411 I have made the required changes. Please review.

File: test/test_torch.py [hunk 4]
```diff
@@ -10013,6 +10015,23 @@ def test_ops(op, string_of_function_name, expected_output):
...
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_cumlogsumexp(self):
+        a = torch.randn(5, 4)
+        a[0, 0] = inf
+        a[1, :] = -inf
+        actual = a.cumlogsumexp(1)
+        max = np.maximum.accumulate(a.numpy(), axis=1)
+        expected = max + np.log(np.cumsum(np.exp(a - max), axis=1))
+        self.assertEqual(a.dtype, actual.dtype)
+        self.assertEqual(expected.shape, actual.shape)
```
Comment (anjali411) [inline test/test_torch.py]: Add a test to check if there's a RuntimeError when the dtype for provided output variable is not the same as input

**Comment by anjali411** [comment]
> @anjali411 I have made the required changes. Please review.

Hey @pandeykartikey thanks for adding the tests! It looks like the autograd test is failing on both cpu and cuda. can you look into it?
Also I was wondering if logcumsumexp would be a better choice for the name of the function?

**Comment by pandeykartikey** [comment]
@anjali411 I can't seem to understand the issue can you help me with that?

**Comment by albanD** [comment]
Hi @pandeykartikey the test is checking if the gradient you compute in `logcumsumexp_backward` match the ones computed with finite differentiation.
You can check manually by using `torch.autograd.gradcheck` with your function and sample inputs (in double precision so that the finite difference is precise enough).
I guess this can be due to three things:
- `logcumsumexp_backward` is not computing the right thing
- `logcumsumexp_backward` is not numerically stable and so the computed gradients are far from the expected ones
- `logcumsumexp` leads to numerical stability issue when doing finite difference.

**Comment by agadetsky** [comment]
@pandeykartikey do you have any updates?

File: tools/autograd/templates/Functions.cpp [hunk 5]
```diff
@@ -430,6 +430,19 @@ Tensor solve_backward_A(const Tensor & grad, const Tensor & self, const Tensor &
   return -at::matmul(grad_self, solution.transpose(-2, -1));
 }
 
+Tensor logcumsumexp_backward(Tensor grad, const Tensor & self, Tensor result, int64_t dim) {
+  grad = grad*(-result).exp();
```
Comment (anjali411) [inline tools/autograd/templates/Functions.cpp]: I don't understand why you have the minus sign here? I think it should be:
```grad = grad*(result.exp());```

File: tools/autograd/templates/Functions.cpp [hunk 6]
```diff
@@ -430,6 +430,19 @@ Tensor solve_backward_A(const Tensor & grad, const Tensor & self, const Tensor &
...
+  if (grad.dim() == 0 || grad.numel() == 0) {
```
Comment (anjali411) [inline tools/autograd/templates/Functions.cpp]: this condition should precede the previous line. it should be the first condition to be checked before we begin any computations

File: tools/autograd/templates/Functions.cpp [hunk 7]
```diff
@@ -430,6 +430,19 @@ Tensor solve_backward_A(const Tensor & grad, const Tensor & self, const Tensor &
...
+    return grad;
+  }
+  grad = grad*(-result).exp();
+  auto ret = at::cumsum(grad, dim);
+  auto ret_sum = ret.narrow(dim, ret.size(dim) - 1, 1).clone(at::MemoryFormat::Preserve);
+  ret -= ret_sum.expand(ret.sizes());
+  ret += grad;
+  return grad*self.exp();
```
Comment (anjali411) [inline tools/autograd/templates/Functions.cpp]: I am not sure what you are doing here. all you need is:
```
ret = cumsum_backward(ret);
ret = ret * self.exp();
return ret;
```

**Comment by anjali411** [comment]
Hi @pandeykartikey I think the backward function is incorrect in its current state. I left some comments on it. let me know if you have any questions.

cc. @albanD @agadetsky

**Comment by agadetsky** [comment]
@anjali411, seems like @pandeykartikey gave up on this, do you have time to finish this PR?

**Comment by anjali411** [comment]
> @anjali411, seems like @pandeykartikey gave up on this, do you have time to finish this PR?

I am not sure if I will be able to get to it this week, but I can try

**Comment by agadetsky** [comment]
> > @anjali411, seems like @pandeykartikey gave up on this, do you have time to finish this PR?
> 
> I am not sure if I will be able to get to it this week, but I can try

thank you very much, will be waiting for your PR!

**Comment by agadetsky** [comment]
@anjali411 do you have any updates on this?

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -1378,6 +1378,25 @@ def merge_dicts(*dicts):
...
 """.format(**common_args))
 
+add_docstr(torch.logcumsumexp,
+           r"""
+logcumsumexp(input, dim, out=None) -> Tensor
+Returns the logarithm of the cumulative summation of the exponentiation of
+elements of :attr:`input` in the dimension :attr:`dim`. 
+
+.. math::
+    \text{{logsumexp}}(x)_{{i}} = \log \sum_j \exp(x_{{ij}})
```
Comment (gchanan) [inline torch/_torch_docs.py]: I'm not an expert on our convention here, but this doesn't seem correct.

For one, this looks like a reduction, but this function isn't a reduction.

I would have expected something roughly like:
` \text{{logsumexp}}(x)_{{ij}} = \log \sum_{k<=j} \exp(x_{{ik}})`

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -1378,6 +1378,25 @@ def merge_dicts(*dicts):
...
+Args:
+    {input}
+    dim  (int): the dimension to do the operation over
+    {out}
+Example::
+    >>> a = torch.randn(10)
+    >>> torch.logcumsumexp(a, dim=0)
+    tensor([-0.42296738, -0.04462666,  0.86278635,  0.94622083,  1.05277811,
+        1.39202815,  1.83525007,  1.84492621,  2.06084887,  2.06844475]))
```
Comment (gchanan) [inline torch/_torch_docs.py]: nit: can you line up this output?

File: test/test_namedtensor.py [hunk 10]
```diff
@@ -988,6 +988,13 @@ def test_ops(op):
         test_ops(torch.cummax)
         test_ops(torch.cummin)
 
+    def test_logcumsumexp(self):
+        for device in torch.testing.get_all_device_types():
+            names = ('N', 'D')
+            tensor = torch.rand(2, 3, names=names)
+            result = torch.logcumsumexp(tensor, 0)
```
Comment (gchanan) [inline test/test_namedtensor.py]: I would have expected passing the dimension name here ('N' or 'D'), not an integer.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 11]
```diff
@@ -180,6 +180,26 @@ static TensorIterator make_reduction(
...
+}
+
+Tensor& logcumsumexp_out(Tensor& result, const Tensor& self, int64_t dim) {
+  check_scalar_type_device_layout_equal(result, self);
+  {
+    NoNamesGuard guard;
+    result.resize_(self.sizes());
+    auto cummax_values = std::get<0>(at::cummax(self, dim));
+    result = at::exp(self - cummax_values);
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: setting the result in this way isn't correct, you need to write into the tensor.

**Comment by agadetsky** [comment]
@anjali411 hi! do you have updates on this? Are you able to finish this PR?

**Comment by agadetsky** [comment]
@anjali411 hello! are you planning to finish this PR?

**Comment by anjali411** [comment]
> @anjali411 hello! are you planning to finish this PR?

Hi @agadetsky I won't be able to get to this PR anytime soon because of other commitments. Feel free to take over, if you or anyone else gets to it before me.

File: torch/_torch_docs.py (hunk 8 above)
Comment (anjali411) [reply torch/_torch_docs.py]: I think following the notation in cumsumexp documentation, we should write:

```
For summation index :math:`j` given by `dim` and other indices :math:`i`, the result is

    .. math::
        \text{{logcumsumexp}}(x)_{{i}} = \log \cumsum_j \exp(x_{{ij}})
```

**Comment by albertfgu** [comment]
Hi, I took a look at the implementation https://github.com/pytorch/pytorch/blob/4cc31d632d5ae21aaa79fb286b96a83bb0645a37/aten/src/ATen/native/ReduceOps.cpp#L189
since I am in need of a logcumsumexp operation.

I believe this implementation is incorrect as written.
- A simple example such as where the input is arange(n) illustrates why 
- Note that the correct python implementation that this is based off (https://github.com/pytorch/pytorch/pull/32876) is a quadratic instead of linear operation, since you have to broadcast the cummax
- I don't see an easy way to implement this in linear time using cumsum and cummax as primitives

**Comment by tridao** [comment]
I think an easier way to implement this is as a prefix scan with a binary operation log_add_exp(x, y) = log(exp(x) + exp(y)). This is linear time (instead of quadratic time with cummax and broadcasting).
Tensorflow implement it exactly this way: https://www.tensorflow.org/api_docs/python/tf/math/cumulative_logsumexp
I think it's not too bad to adapt the cumsum implementation to do this, since the inclusive_scan machinery is there.

**Comment by kshitij12345** [comment]
@anjali411 @agadetsky  I would like to take over the PR.

**Comment by mruberry** [comment]
Good news is that we now have logcumsumexp!

**Pull request closed without merging**